File: leetcode/46-Permutations/permutations.py
# Given a collection of distinct numbers, return all possible permutations.
#
# For example,
# [1,2,3] have the following permutations:
# [
#   [1,2,3],
#   [1,3,2],
#   [2,1,3],
#   [2,3,1],
#   [3,1,2],
#   [3,2,1]
# ]

class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def dfs(nums, path, visited, res):
            if len(path) == len(nums):
                res.append(path)
                return
            for i in range(len(nums)):
                if visited[i] == 0:
                    visited[i] = 1
                    dfs(nums, path+[nums[i]], visited, res)
                    visited[i] = 0
        res = []
        visited = [0] * len(nums)
        dfs(nums, [], visited, res)
        return res

if __name__ == "__main__":
    sol = Solution()
    print(sol.permute([1,2,3]))

File: leetcode/46-Permutations/go/permutations.go
package permutations

// Given a collection of distinct integers, return all possible permutations.

// Example:

// Input: [1,2,3]
// Output:
// [
//   [1,2,3],
//   [1,3,2],
//   [2,1,3],
//   [2,3,1],
//   [3,1,2],
//   [3,2,1]
// ]

func permute(nums []int) [][]int {
	var tmp []int
	var res [][]int
	invalid_pos := make([]bool, len(nums))
	res = backtrack(nums, tmp, invalid_pos, res)
	return res
}

// backtrack generates all the permutations of the given `nums` and put them into `res`.
// `invalid_pos` keeps track of which number is valid to pick from `nums` to form a permutation
// `tmp`
func backtrack(nums []int, tmp []int, invalid_pos []bool, res [][]int) [][]int {
	if len(tmp) == len(nums) {
		tmp2 := make([]int, len(tmp))
		copy(tmp2, tmp)
		res = append(res, tmp2)
	} else {
		for i, num := range nums {
			if invalid_pos[i] {
				continue
			}
			invalid_pos[i] = true
			tmp = append(tmp, num)
			res = backtrack(nums, tmp, invalid_pos, res)
			invalid_pos[i] = false
			tmp = tmp[:len(tmp)-1]
		}
	}
	return res
}
